src/Store/VimStoreConnector.re

```diff
@@ -317,17 +319,59 @@ let start =
       open Vim.BufferUpdate;
       Log.info("Vim - Buffer update: " ++ string_of_int(update.id));
       open Core.Types;
+      open Model.State;
+
+      let isFull = update.endLine == (-1);
+
+      let bufferMaybe =
+        Some(getState())
+        |> Option.map(state => state.buffers)
+        |> Option.bind(Model.Buffers.getBuffer(update.id));
+
+      // If this is a 'full' update, check if there was a previous buffer.
+      // We need to keep track of the previous line count for some
+      // buffer synchronization strategies (ie, extension host)
+      let endLine =
+        if (isFull) {
+          bufferMaybe
+          |> Option.map(b => Model.Buffer.getNumberOfLines(b) + 1)
+          |> Option.value(~default=update.startLine);
+        } else {
+          update.endLine;
+        };
+
       let bu =
-        Core.Types.BufferUpdate.create(
+        Core.BufferUpdate.create(
           ~id=update.id,
+          ~isFull,
           ~startLine=Index.OneBasedIndex(update.startLine),
-          ~endLine=Index.OneBasedIndex(update.endLine),
+          ~endLine=Index.OneBasedIndex(endLine),
           ~lines=update.lines,
           ~version=update.version,
           (),
         );
 
-      dispatch(Model.Actions.BufferUpdate(bu));
+      // We need to filter out updates that come 'out-of-order'. This shouldn't need to be done
+      // at this layer - but there are some bugs with the updates we get from 'reason-libvim'.
+      // In particular, with undo / redo, we get multiple updates - a full update (version+2) and
+      // an incremental update (version+1) _after_ the full update. If we apply the incremental update
+      // after the full update, ignoring the version, we'll get incorrect results.
+      //
+      // The fix really belongs in reason-libvim - we should always be trust the order we get from the updates,
+      // and any of this filtering or manipulation of updates should be handled and tested there.
+      let shouldApply =
+        bufferMaybe
+        |> Option.map(Model.Buffer.shouldApplyUpdate(bu))
+        |> Option.value(~default=true);
```

Comment:
This feels a bit convoluted. How about `Option.map(Model.Buffer.shouldApplyUpdate(bu), bufferMaybe) != Some(false)`. Isn't that at least a little bit clearer?

Also, just opening `Oni_Model` at the top would remove quite a bit of noise. opening modules that only contain module aliases is usually fine.

Reply:
Sounds good - I think the confusing aspect of the logic is that you still have to think about the `None` case. But it is cleaner. And hopefully the confusing part we can go away by fixing the root-cause bug in reason-libvim.